kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/editing/SkaffoldSchemaFileProvider.kt

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2018 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.container.tools.skaffold.editing
+
+import com.google.container.tools.skaffold.SkaffoldFileService
+import com.intellij.openapi.vfs.VirtualFile
+import com.jetbrains.jsonSchema.extension.JsonSchemaFileProvider
+import com.jetbrains.jsonSchema.extension.JsonSchemaProviderFactory
+import com.jetbrains.jsonSchema.extension.SchemaType
+
+const val SKAFFOLD_SCHEMA_NAME = "skaffold"
+
+/**
+ * A [JsonSchemaFileProvider] that attaches a bundled JSON schema to the editor corresponding to the
+ * version of Skaffold.
+ */
+class SkaffoldSchemaFileProvider(private val skaffoldSchemaVersion: String)
+    : JsonSchemaFileProvider {
+
+    override fun getName(): String = SKAFFOLD_SCHEMA_NAME
+
+    /**
+     * Returns true if the currently open file is a valid Skaffold file, and the Skaffold version
+     * associated with this schema provider matches the Skaffold version in the currently open file.
+     */
+    override fun isAvailable(file: VirtualFile): Boolean {
+        val skaffoldVersion: String? = SkaffoldFileService.instance.getSkaffoldVersion(file)
+
+        return SkaffoldFileService.instance.isSkaffoldFile(file) &&
+                skaffoldVersion == skaffoldSchemaVersion
+    }
+
+    override fun getSchemaFile(): VirtualFile? {
+        return JsonSchemaProviderFactory
+                .getResourceFile(this::class.java, "/schemas/$skaffoldSchemaVersion.json")
```

Comment:
Convenient method! Is this the one used for other schemas too?

Reply:
currently this is only used to load the Skaffold schemas. I'm not sure if that is what you were thinking of though?

Reply:
Yeah, I thought it's convenient to just convert class resource into a virtual file, they need to make it more broad eventually. Just interested if Jetbrains code uses it as well for schema loading.

Reply:
yeah this is how the platform does this as well